doc: Correct uint256 hex string endianness

hodlinator opened the pull request:
This PR is a follow-up to #30436.

Only changes test-code and modifies/adds comments.

Byte order of hex string representation was wrongfully documented as little-endian, but are in fact closer to "big-endian" (endianness is a memory-order concept rather than a numeric concept). `[arith_]uint256` both store their data in arrays with little-endian byte order (`arith_uint256` has host byte order within each `uint32_t` element).

**uint256_tests.cpp** - Avoid using variable from the left side of the condition in the right side. Credits to @maflcko: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1688273553

**setup_common.cpp** - Skip needless ArithToUint256-conversion. Credits to @stickies-v: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1688621638

---

<details>
<summary>

## Logical reasoning for endianness

</summary>

1. Comparing an `arith_uint256` (`base_uint<256>`) to a `uint64_t` compares the beginning of the array, and verifies the remaining elements are zero.
```C++
template <unsigned int BITS>
bool base_uint<BITS>::EqualTo(uint64_t b) const
{
    for (int i = WIDTH - 1; i >= 2; i--) {
        if (pn[i])
            return false;
    }
    if (pn[1] != (b >> 32))
        return false;
    if (pn[0] != (b & 0xfffffffful))
        return false;
    return true;
}
```
...that is consistent with little endian ordering of the array.

2. They have the same endianness (but `arith_*` has host-ordering of each `uint32_t` element):
```C++
arith_uint256 UintToArith256(const uint256 &a)
{
    arith_uint256 b;
    for(int x=0; x<b.WIDTH; ++x)
        b.pn[x] = ReadLE32(a.begin() + x*4);
    return b;
}
```

### String conversions

The reversal of order which happens when converting hex-strings <=> uint256 means strings are actually closer to big-endian, see the end of `base_blob<BITS>::SetHexDeprecated`:
```C++
    unsigned char* p1 = m_data.data();
    unsigned char* pend = p1 + WIDTH;
    while (digits > 0 && p1 < pend) {
        *p1 = ::HexDigit(trimmed[--digits]);
        if (digits > 0) {
            *p1 |= ((unsigned char)::HexDigit(trimmed[--digits]) << 4);
            p1++;
        }
    }
```
Same reversal here:
```C++
template <unsigned int BITS>
std::string base_blob<BITS>::GetHex() const
{
    uint8_t m_data_rev[WIDTH];
    for (int i = 0; i < WIDTH; ++i) {
        m_data_rev[i] = m_data[WIDTH - 1 - i];
    }
    return HexStr(m_data_rev);
}
```
It now makes sense to me that `SetHexDeprecated`, upon receiving a shorter hex string that requires zero-padding, would pad as if the missing hex chars where towards the end of the little-endian byte array, as they are the most significant bytes. "Big-endian" string representation is also consistent with the case where `SetHexDeprecated` receives too many hex digits and discards the leftmost ones, as a form of integer narrowing takes place.

### How I got it wrong in #30436

Previously I used the less than (`<`) comparison to prove endianness, but for `uint256` it uses `memcmp` and thereby gives priority to the *lower* bytes at the beginning of the array.
```C++
    constexpr int Compare(const base_blob& other) const { return std::memcmp(m_data.data(), other.m_data.data(), WIDTH); }
```

`arith_uint256` is different in that it begins by comparing the bytes from the end, as it is using little endian representation, where the bytes toward the end are more significant.
```C++
template <unsigned int BITS>
int base_uint<BITS>::CompareTo(const base_uint<BITS>& b) const
{
    for (int i = WIDTH - 1; i >= 0; i--) {
        if (pn[i] < b.pn[i])
            return -1;
        if (pn[i] > b.pn[i])
            return 1;
    }
    return 0;
}
```
(The commit documents that `base_blob::Compare()` is doing lexicographic ordering unlike the `arith_*`-variant which is doing numeric ordering).

</details>

paplorinc commented:
Thanks for your detailed reasoning, @sipa!

Documenting the endianness was my idea, and I appreciate your insights. I will investigate your claims in more detail later as I am currently traveling.

paplorinc commented:
After further consideration and research, I believe I've identified a key point that may help clarify our discussion about endianness in the codebase. I'd like to propose a more nuanced view of the situation.
We seem to agree that the `base16` to `base256` conversion can indeed have an endianness.
Based on the Wikipedia definition of endianness as "the order in which bytes within a word of digital data are transmitted over a data communication medium or addressed (by rising addresses) in computer memory, counting only **byte significance compared to earliness**", the resulting bytes (representing any data) can also be stored forward or backward, which aligns with this concept of significance compared to earliness. So according to this, arrays can also have endianness.

Currently, it seems we have a form of double endian encoding in our codebase, where:
* The base16 to base256 conversion has an endianness.
* The storage of the resulting bytes can also reflect endianness.

This double encoding might be the root of why we viewed this differently.

While I agree that your proposed approach could lead to more precise terminology here, upon digging into the codebase, it appears that endianness terms are used more broadly than your strict interpretation would suggest:

**Endianness terminology used for hexadecimal representations**:
* Transaction IDs: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L620
* Hashes: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L621

**Endianness applied to non-numeric data**:
* Base58 encoding: https://github.com/bitcoin/bitcoin/blob/master/src/base58.cpp#L98
* CRC32: https://github.com/bitcoin/bitcoin/blob/master/src/crc32c/src/crc32c_read_le.h#L17
* SipHash: https://github.com/bitcoin/bitcoin/blob/master/src/crypto/siphash.h#L25

**uint256 treated as having specific endianness**:
* In policy code: https://github.com/bitcoin/bitcoin/blob/master/src/policy/packages.h#L93

It also doesn't help that the `uint256` getters themselves reverse the byte order before returning it (https://github.com/bitcoin/bitcoin/blob/master/src/uint256.cpp#L11-L18) and that reading `uint64` chunks even considers the native endianness https://github.com/bitcoin/bitcoin/blob/master/src/uint256.h#L79.

The problem doesn't seem to be as simple as just changing the terminology, it seems to be deeply ingrained in the codebase. I assume that we can't actually simplify the code in most places (maybe for SipHash-type usages where the result is local and temporary), so we have to either refactor or document, right? How do you suggest we tackle this?

hodlinator commented:
Thank you for your feedback @paplorinc & @sipa!

I am convinced by sipa's argument that endianness i primarily an in-memory representation order. Numbers like 4607 (= 0x11FF) do not have an endianness to them, neither should strings containing them, even though it is very tempting for hexadecimal.

07de45a882d43fd5567b2a6b7c8a2da3d6dbda3d contains much of sipa's input. Not sure what to do with pre-existing code @paplorinc found that uses endianness terminology imprecisely.

> Transaction IDs: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L620

Specifies txid as "transaction id encoded in little-endian hexadecimal".

I dug into the behavior of the `HashWriter`:

```diff
diff --git a/src/test/hash_tests.cpp b/src/test/hash_tests.cpp
index 51f1d4c840..590af57ff6 100644
--- a/src/test/hash_tests.cpp
+++ b/src/test/hash_tests.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include "uint256.h"
 #include <clientversion.h>
 #include <crypto/siphash.h>
 #include <hash.h>
@@ -146,6 +147,10 @@ BOOST_AUTO_TEST_CASE(siphash)
         BOOST_CHECK_EQUAL(SipHashUint256(k1, k2, x), sip256.Finalize());
         BOOST_CHECK_EQUAL(SipHashUint256Extra(k1, k2, x, n), sip288.Finalize());
     }
+
+    HashWriter writer{};
+    writer.write(std::vector<std::byte>{std::byte{'1'}, std::byte{'2'}, std::byte{'3'}});
+    BOOST_CHECK_EQUAL(writer.GetSHA256(), uint256S("a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3"));
 }
 
 BOOST_AUTO_TEST_SUITE_END()
```
The "a665a45.."-string was provided by 2 online SHA256 calculators.

Result:
```
test/hash_tests.cpp(153): error: in "hash_tests/siphash": check writer.GetSHA256() == uint256S("a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3") has failed [e37aa2f7f7868e997ea01fff3f1f4aa0b84fdcef67487e419d2f422059a465a6 != a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3]
```

Conclusion: Byte order is unnecessarily reversed in this case, but now hard to change. Should probably be documented as "transaction id encoded in reverse-order hexadecimal" but this PR is just concerned with shortcomings of #30436.

ryanofsky commented on src/uint256.h:
```diff
@@ -53,15 +53,27 @@ class base_blob
 
-    // Hex string representations are little-endian.
+    /* Hex string representations in reverse byte order of the internal array.
+     * This is consistent with the internal array of arith_* being little-endian.
```
In commit "doc + test: Correct uint256 hex string endianness" (07de45a882d43fd5567b2a6b7c8a2da3d6dbda3d)

Statement "This is consistent with the internal array of arith_* being little-endian." seems a bit vague to me. Also calling internal representation of the arithmetic class little endian might not be ideal because endianness normally refers to bytes and the arithmetic class stores an array of uint32_t numbers, not of bytes.

I think I would suggest changing line to:

- The reverse hex string representation can be a convenient way of showing the blob as a number, because it is consistent with the way the `base_uint` class converts blobs to numbers.

  `base_uint` treats the blob as an array of bytes with the least significant byte first, and most significant byte last. Because numbers are typically written with the most significant digit first, and least significant digit last, the reverse hex display of the blob corresponds to the same arithmetic number `base_uint` interprets from the blob.

paplorinc replied:
👍, I like that we're not saying endianness for arrays anymore (as suggested by @sipa as well), but seems we still do here. Would be nice if we could eventually unify throughout the codebase.

ryanofsky reviewed (approved):
Code review ACK 07de45a882d43fd5567b2a6b7c8a2da3d6dbda3d. I do think this is an improvement over the status quo, but I think a lot of things were less clear than they could be so I left a lot of suggestions.

paplorinc commented on src/uint256.h:
```diff
@@ -53,15 +53,27 @@ class base_blob
-    /** Unlike FromHex this accepts any invalid input, thus it is fragile and deprecated */
+    /** Unlike FromHex this accepts any invalid input, thus it is fragile and deprecated!
+     *
+     * Hex strings that don't specify enough bytes to fill the internal array
```
nit: the hex strings don't actually provide bytes, the bytes are computed, right?

paplorinc commented on src/uint256.h:
```diff
@@ -93,7 +105,7 @@ class base_blob
 namespace detail {
 /**
- * Writes the hex string (treated as little-endian) into a new uintN_t object
+ * Writes the hex string (in reverse byte order) into a new uintN_t object
```
👍 I like this wording a lot more

paplorinc commented on src/uint256.h:
```diff
@@ -53,15 +53,27 @@ class base_blob
     }
 
+    /** Lexicographic ordering
+     * @note Does NOT match the ordering on the corresponding \ref
```
this is really messy, I wonder if we have a chance to untangle any of this - or whether documenting it is our only possible solution

paplorinc commented on src/test/uint256_tests.cpp:
```diff
@@ -152,9 +153,15 @@ BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
+    // Ensure last element comparisons give a different result (swapped params):
+    BOOST_CHECK_LT(*(R1L.end()-1), *(R2L.end()-1));
+    // Hex string representation is reversed, before lexicographic ordering:
+    BOOST_CHECK_LT(uint256S("1000000000000000000000000000000000000000000000000000000000000000"),
+                   uint256S("0000000000000000000000000000000000000000000000000000000000000001"));
```
Good observations

paplorinc commented on src/arith_uint256.h:
```diff
@@ -196,6 +197,7 @@ class base_uint
         return ret;
     }
 
+    /** Numeric ordering (unlike \ref base_blob::Compare) */
```
no strong preference either way, just noticing that in other cases it's rather called https://en.wikipedia.org/wiki/Natural_sort_order

l0rinc reviewed (approved):
utACK 07de45a882d43fd5567b2a6b7c8a2da3d6dbda3d

I love that we're striving to making the campground cleaner than we found it, we need more of these kind of changes <3
Agree with @ryanofsky, some of the comments could still use some clarification, but I'm fine with it as-is as well.

hodlinator committed:
doc + test: Correct uint256 hex string endianness

Follow-up to #30436.

uint256 string representation was wrongfully documented as little-endian due to them being reversed by GetHex() etc, and base_blob::Compare() giving most significance to the beginning of the internal array. They are closer to "big-endian", but this commit tries to be even more precise than that.

uint256_tests.cpp - Avoid using variable from the left side of the condition in the right side.

setup_common.cpp - Skip needless ArithToUint256-conversion.

hodlinator commented:
Latest force push to fix CI failure: https://github.com/bitcoin/bitcoin/pull/30526/checks?check_run_id=28237349796

paplorinc commented on src/uint256.h:
```diff
@@ -53,17 +53,46 @@ class base_blob
+     */
     void SetHexDeprecated(std::string_view str);
     std::string ToString() const;
+    /**@}*/
```
was this deliberate? if so, what does it mean?

hodlinator replied:
https://www.doxygen.nl/manual/grouping.html#memgroup

Check out `make -j<cores> docs`.

paplorinc replied:
hah, haven't seen that yet, thanks

ryanofsky commented on src/arith_uint256.h:
```diff
@@ -26,6 +26,7 @@ class base_uint
 protected:
     static_assert(BITS / 32 > 0 && BITS % 32 == 0, "Template parameter BITS must be a positive multiple of 32.");
     static constexpr int WIDTH = BITS / 32;
+    /** Fully little-endian byte ordering on little-endian architectures. */
```
> > Maybe just say number is stored in the array with least significant bits first and most significant bits last.
> 
> That is incorrect. The least significant _byte_ is stored first for `uint32_t` on the little-endian archs. If one wants to avoid mentioning architecture, one could possibly say "the number is stored in an array with the least significant uint32_t chunk first".

It is correct, it's just saying the highest bits of the number are in the last array element and the lowest bits of the number are in the first array element. But you could definitely say chunks instead of bits if you prefer that.

> > The current comment describing byte ordering on one specific architecture seems like an indirect way of describing the array when array elements aren't bytes, and the class itself is architecture independent.
> 
> Do you mean mentioning architecture on a protected member instills fear that the users of the class will be exposed to architecture dependent behavior?

It's just confusing and indirect.

hodlinator replied:
> It is correct, it's just saying the highest bits of the number are in the last array element and the lowest bits of the number are in the first array element. But you could definitely say chunks instead of bits if you prefer that.

It is correct if by "bits" you mean 32-bit chunks together I guess, then the lowest is at the beginning and the highest at the end. I was somehow wrongly interpreting you as meaning *the* highest bit (singular) at the end + *the* lowest bit at the beginning.

> It's just confusing and indirect.

Would this feel less confusing?
```C++
    /** Number is stored as fully little-endian on little-endian archs. */
    uint32_t pn[WIDTH];
```
..? Then it at least specifies what is being stored (but it's the only data-member of `base_uint` so could be considered overly verbose IMO). Kept it short to stay within 80 char line.

ryanofsky reviewed (approved):
Code review ACK 0828a3601c7f323f815598a41dd7854ef25265a7. Thanks for considering the suggestions, and I like the way you now grouped the reverse-hex functions together and added an overall description of the representation.

I do think it would be better if the base_uint class try to didn't try to document the array by describing its memory layout on little endian machines. I think it would be clearer if it just said what order the array elements are in.

ryanofsky commented on src/test/arith_uint256_tests.cpp:
```diff
@@ -278,6 +281,12 @@ BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
+
+    BOOST_CHECK_LT(ZeroL,
+                   OneL);
+    // Verify hex number representation has the most significant bytes first.
```
re: https://github.com/bitcoin/bitcoin/pull/30526#discussion_r1699168468

Still think it would be better to say digits instead of bytes here because this is referring to a number string, not a byte array.

re: https://github.com/bitcoin/bitcoin/pull/30526#discussion_r1700191168

> > When there's a hex encoding, "byte" could refer either to a byte of the hex string or to a byte of the data it represents, and I think you mean the former (because the digits in the two strings are what's reversed, not the bytes).
> 
> I mean that `FF` represents a byte will all bits set, so I guess the latter. The individual hex digits are NOT reversed, `0F` is not interpreted as `F0`, it is the order of the _bytes_ that is reversed, `12EF` => `EF12`.

I don't understand why you are thinking about bytes at all here when this is a number string. When I am saying the digits are reversed I am saying "12345" is becoming "54321". I am not saying "12345" is becoming "103254". In your test you making sure the number "0000000000000000000000000000000000000000000000000000000000000001" is less than the number "1000000000000000000000000000000000000000000000000000000000000000". The digits in these numbers are reversed, not the bytes. The second number is straightforwardly higher than the first because the highest digit is bigger (1 > 0). It is true you can also say that the highest byte is bigger (10 > 0), but it's an odd thing to say because this is just a number and there should be no need to think about bytes at all to make sure a numeric comparison is working correctly.

paplorinc commented:
ACK 0828a3601c7f323f815598a41dd7854ef25265a7

l0rinc reviewed (approved):
ACK 73e3fa10b4dd63b7767d6b6f270df66971067341

ryanofsky reviewed (approved):
Code review ACK 73e3fa10b4dd63b7767d6b6f270df66971067341

Just tweaked code comments since last review. I think this can be merged even though from the review comments it seems like there could be more more followups later. Everything here now seems like an improvement over the status quo.

ryanofsky reviewed (commented):
Merged with only 2 acks, since this is a documentation & test change.

It does seem like there are some possible followups here:

1. Providing more context for reverse-byte hex representation https://github.com/bitcoin/bitcoin/pull/30526#discussion_r1703467650
2. Improving RPC documentation as suggested by paplorinc in https://github.com/bitcoin/bitcoin/pull/30526#issuecomment-2254461008
3. Documenting base_uint::pn array https://github.com/bitcoin/bitcoin/pull/30526#discussion_r1699163923
4. Maybe adding stricter tests for comparison functions https://github.com/bitcoin/bitcoin/pull/30526#discussion_r1699168796 and https://github.com/bitcoin/bitcoin/pull/30526#discussion_r1699169882

---

re: https://github.com/bitcoin/bitcoin/pull/30526#issuecomment-2254461008

> While I agree that your proposed approach could lead to more precise terminology here, upon digging into the codebase, it appears that endianness terms are used more broadly than your strict interpretation would suggest:
> 
> **Endianness terminology used for hexadecimal representations**:
> 
> * Transaction IDs: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L620
> * Hashes: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L621

These are good finds and I think hodlinator's suggestion to document these as "transaction id encoded in reverse-order hexadecimal" could an improvement, although maybe it is ambiguous. "Reverse-order hexadecimal" sounds like the hexadecimal string might be reversed, rather than the bytes represented by the hexadecimal string being reversed. I might suggest changing documentation of these fields to:

- txid hash with bytes reversed, encoded in hex
- wtxid hash with bytes reversed, encoded in hex

but maybe something else would be better.

> **Endianness applied to non-numeric data**:
> 
> * Base58 encoding: https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/base58.cpp#L98

I think this comment is using "big-endian" correctly, though maybe not very clearly. The idea behind base58 encoding is that you take a byte array, interpret the byte array as a big-endian number, and display that number in base-58 (with a special prefix of 1's if the original byte array began with 0's). (The [python](https://github.com/bitcoin/bitcoin/blob/1a7d20509f46f0cd38302da78dba9a0c9bb24226/test/functional/test_framework/address.py#L70) version of this function is much easier to understand, for reference.)

> * CRC32: https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/crc32c/src/crc32c_read_le.h#L17

This seems like a totally correct comment. This function is taking bytes and converting them to an integer. Seems like a straightforward case of using "endianness" to describe byte order of numeric data.

> * SipHash: https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/crypto/siphash.h#L25

This also seems correct. The siphash class was two Write methods. There is a Write method take bytes as input, and another Write method that takes 64 bit numbers as input. The documentation is just saying if you give that one a number, the little endian representation of the number is what is hashed.

> **uint256 treated as having specific endianness**:
> 
> * In policy code: https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/policy/packages.h#L93

This seems like a borderline case and probably would be good to clarify. It's just describing the way the wtxids are sorted before they are hashed. Instead of the implementation sorting the wtxid so they are in lexigraphic order, it sorts them so that the reversed bytes of the wtxid's would be in lexicographic order. This is equivalent to treating the wtxid bytes as little endian numbers and sorting the numbers. 

> It also doesn't help that the `uint256` getters (and testers like `ArrayToString`) themselves reverse the byte order before returning it (https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/uint256.cpp#L11-L18)

This isn't great but I think it's baked in and the best thing we can do is document it.

> and that reading `uint64` chunks even considers the native endianness https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/uint256.h#L79.

This should be fine, because it is only using native endianness as an optimization, and does the same thing on big and little endian systems. It is just interpreting bytes of the blob as a little endian number, and on little endian systems this can be done efficiently with memcpy.

> The problem doesn't seem to be as simple as just changing the terminology, it seems to be deeply ingrained in the codebase. I assume that we can't actually simplify the code in most places (maybe for SipHash-type usages where the result is local and temporary), so we have to either refactor or document, right? How do you suggest we tackle this?

I think we can clean up documentation various places, name functions better, and maybe in the long term move towards sipa' suggestion to replace these classes with `using uint256 = std::array<std::byte, 32>`.
